### PR TITLE
[Compiler-v2] Fix freeze_ref

### LINKED ISSUE #13927: [Bug][Compiler-v2] FREEZEREF_EXISTS_MUTABLE_BORROW_ERROR
# 🐛 Bug

When compiling the following code with V2, the bytecode verification error `FREEZEREF_EXISTS_MUTABLE_BORROW_ERROR` is raised

```
module 0x8675309::Tester {

    fun foo(_other: &mut u64): &mut u64 {
        _other
    }

    fun test(result: &mut u64): &u64 {
        let returned_ref = foo(result);
        freeze(result);
        returned_ref
    }
}
```

### PR DESCRIPTION
**Pull request by rahxephon89**
## Description

This PR closes #13927 by adding a condition in `freeze_ref` such that when a mutable reference is being frozen, if there exist mutable borrows from it, an error will be raised.

## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?

1) Existing tests pass;
2) Added a new test case `freeze_mutable`.

## Key Areas to Review

Whether the added condition is correct.

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: third_party/move/move-compiler-v2/tests/reference-safety/v1-tests/freeze_combo_invalid.exp [hunk 1]
```diff
@@ -1,5 +1,21 @@
 
 Diagnostics:
+error: cannot freeze local `s` since other mutable usages for this reference exist
```
Comment (rahxephon89) [inline third_party/move/move-compiler-v2/tests/reference-safety/v1-tests/freeze_combo_invalid.exp:3]: V1 has this error but was missed previously in V2.

File: third_party/move/move-compiler-v2/tests/reference-safety/v1-tests/freeze_field_invalid.exp [hunk 2]
```diff
@@ -1,5 +1,13 @@
 
 Diagnostics:
+error: cannot freeze local `s` since other mutable usages for this reference exist
```
Comment (rahxephon89) [inline third_party/move/move-compiler-v2/tests/reference-safety/v1-tests/freeze_field_invalid.exp:3]: V1 has this error but was missed previously in V2.

File: third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs [hunk 3]
```diff
@@ -1793,7 +1793,9 @@ impl<'env, 'state> LifetimeAnalysisStep<'env, 'state> {
                 {
                     continue;
                 }
-                if other_label == label {
+                if other_label == label
```
Comment (rahxephon89) [inline third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs:1796]: apart from the same memory location, locations mutably borrowed from `label` are also included.

**Review by wrwg — APPROVED** [review]
Nice, seems to be in fact a buug fix on the heuristic part (b) trying to simulate v1 overrestrictive behavior.

File: third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs (hunk 3 above)
Comment (vineethk) [reply third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs:1822]: Could we add this as a comment here, please?

File: third_party/move/move-compiler-v2/tests/reference-safety/freeze_mutable.move [hunk 4]
```diff
@@ -0,0 +1,12 @@
+module 0x8675309::Tester {
+
+    fun foo(_other: &mut u64): &mut u64 {
+        _other
+    }
+
+    fun test(result: &mut u64): &u64 {
+        let returned_ref = foo(result);
+        freeze(result);
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/tests/reference-safety/freeze_mutable.move:9]: Could we add an example where `returned_ref` is not used later, like here? In such a case, we should not give an error, right?

**Review by vineethk — APPROVED** [review]

File: third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs (hunk 3 above)
Comment (vineethk) [reply third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs:1822]: @rahxephon89 Just double checking if this comment was addressed.

File: third_party/move/move-compiler-v2/tests/reference-safety/freeze_mutable.move (hunk 4 above)
Comment (vineethk) [reply third_party/move/move-compiler-v2/tests/reference-safety/freeze_mutable.move:9]: @rahxephon89 Just checking if this comment was addressed.

Comment (rahxephon89) [reply third_party/move/move-compiler-v2/tests/reference-safety/freeze_mutable.move:9]: The PR was unfortunately automatically merged but this PR shows that no error is generated: https://github.com/aptos-labs/aptos-core/pull/14248

**Pull request merged**